be/src/exec/pipeline/runtime_filter_types.h

```diff
@@ -0,0 +1,336 @@
+// and every HashJoinBuildOperatorFactory has its corresponding RuntimeFilterCollector.
+struct RuntimeFilterCollector {
+    RuntimeFilterCollector(RuntimeInFilters&& in_filters, RuntimeBloomFilters&& bloom_filters)
+            : _in_filters(std::move(in_filters)), _bloom_filters(std::move(bloom_filters)) {}
+
+    RuntimeBloomFilters& get_bloom_filters() { return _bloom_filters; }
+    RuntimeInFilters& get_in_filters() { return _in_filters; }
+
+    std::vector<RuntimeInFilterPtr> get_in_filters_bounded_by_tuple_ids(const std::vector<TupleId>& tuple_ids) {
+        std::vector<ExprContext*> selected_in_filters;
+        for (auto* in_filter : _in_filters) {
+            if (in_filter->root()->is_bound(tuple_ids)) {
+                selected_in_filters.push_back(in_filter);
+            }
+        }
+        return selected_in_filters;
+    }
+
+private:
+    // local runtime in-filter
+    RuntimeInFilters _in_filters;
+    // global/local runtime bloom-filter(including max-min filter)
+    RuntimeBloomFilters _bloom_filters;
+};
+
+class RuntimeFilterHolder {
+public:
+    void add_collector(RuntimeFilterCollectorPtr&& collector) {
+        _collector_ownership = std::move(collector);
+        _collector.store(_collector_ownership.get(), std::memory_order_release);
+    }
+    RuntimeFilterCollector* get_collector() { return _collector.load(std::memory_order_acquire); }
+    bool is_ready() { return get_collector() != nullptr; }
+
+private:
+    RuntimeFilterCollectorPtr _collector_ownership;
+    std::atomic<RuntimeFilterCollector*> _collector;
+};
+
+// RuntimeFilterHub is a mediator that used to gather all runtime filters generated by HashJoinBuildOperator instances.
+// It has a RuntimeFilterHolder for each HashJoinBuilder instance, when total runtime filter is generated, then it is
+// added into RuntimeFilterHub; the operators consuming runtime filters inspect RuntimeFilterHub and find out its bounded
+// runtime filters. RuntimeFilterHub is reserved beforehand, and there is no need to use mutex to guard concurrent access.
+class RuntimeFilterHub {
+public:
+    void add_holder(TPlanNodeId id) { _holders.emplace(std::make_pair(id, std::make_unique<RuntimeFilterHolder>())); }
+    void add_collector(TPlanNodeId id, RuntimeFilterCollectorPtr&& collector) {
+        get_holder(id)->add_collector(std::move(collector));
+    }
+
+    RuntimeBloomFilters& get_bloom_filters(TPlanNodeId id) {
+        return get_holder(id)->get_collector()->get_bloom_filters();
+    }
+
+    void close_all_in_filters(RuntimeState* state) {
+        for (auto& [_, holder] : _holders) {
+            if (auto* collector = holder->get_collector()) {
+                for (auto& in_filter : collector->get_in_filters()) {
+                    in_filter->close(state);
+                }
+            }
+        }
+    }
+
+    std::vector<RuntimeFilterHolder*> gather_holders(std::set<TPlanNodeId> ids) {
```

Comment:
The type of `ids` could be `const std::set<TPlanNodeId>&`.

Reply:
ok